### PR TITLE
`ProtostarFixture`

### PR DESCRIPTION
**Pull request by Radinyn**
edited by: @kasperski95 

- added boundary/learning tests to `GatewayFacade`
- added `ProtostarFixture` utility based on various experiments in order to:
  - simplify process of creating integration tests
  - improve tests readability by:
    - removing implementation details from tests
    - improving collocation (cairo files can be defined in Python test suite)
  - reduce the maintainance cost by:
    - preventing changes in interfaces from propagate to all tests
    - allowing removing compilation output fixtures from the codebase  
  - test arguments in integration tests rather than e2e 
- fixed saving Migrator output when cwd was different from the project root

### CODE REVIEW

File: protostar/starknet_gateway/gateway_facade.py [hunk 1]
```diff
@@ -183,6 +186,65 @@ async def declare(
...
+        contract_function: ContractFunction,
+        inputs: Optional[GatewayFacadeSupportedInputType] = None,
+    ):
+        if inputs is None:
+            inputs = {}
+        if isinstance(inputs, List):
+            return await contract_function.call(*inputs)
+        return await contract_function.call(**inputs)
+
+    async def invoke(
```
Comment (Radinyn) [inline protostar/starknet_gateway/gateway_facade.py:233]: leftover from previous PR version

**Review by mkaput — COMMENTED** [review]
could we please split this PR into `migration call` and `ProtostarFixture`? Lots of stuff going on here :(

**Review by mkaput — CHANGES_REQUESTED** [review]

File: protostar/starknet_gateway/gateway_facade.py [hunk 2]
```diff
@@ -19,6 +18,8 @@
 from protostar.starknet_gateway.starknet_request import StarknetRequest
 from protostar.utils.log_color_provider import LogColorProvider
 
+GatewayFacadeSupportedInputType = Union[List[int], Dict[str, Any]]
```
Comment (mkaput) [inline protostar/starknet_gateway/gateway_facade.py:21]: I dont quite understand what this tyle describes from its name

File: protostar/starknet_gateway/starknet_request.py [hunk 3]
```diff
@@ -8,7 +8,7 @@
 
 @dataclass
 class StarknetRequest:
-    Action = Literal["DEPLOY", "DECLARE"]
+    Action = Literal["DEPLOY", "DECLARE", "CALL"]
```
Comment (mkaput) [inline protostar/starknet_gateway/starknet_request.py:11]: Is this in scope of this PR?

**Review by mkaput — CHANGES_REQUESTED** [review]

File: protostar/commands/init/init_command.py [hunk 4]
```diff
@@ -17,11 +18,13 @@ def __init__(
         requester: InputRequester,
         new_project_creator: NewProjectCreator,
         adapted_project_creator: AdaptedProjectCreator,
+        cwd: Path,
     ) -> None:
         super().__init__()
         self._adapted_project_creator = adapted_project_creator
         self._new_project_creator = new_project_creator
         self._requester = requester
+        self._cwd = cwd
```
Comment (mkaput) [inline protostar/commands/init/init_command.py:27]: I don't understand why this code is passing `cwd` parameter everywhere. Wouldn't setting global cwd upon Protostar starting be more fault proof? As Protostar developer, I would expect cwd to be set to project root entire Protostar execution time.

File: tests/integration/gateway/gateway_facade_test.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+@pytest.fixture(name="compiled_contract_path")
+def compiled_contract_path_fixture(protostar: ProtostarFixture) -> Path:
+    return protostar.project_root_path / "build" / "main.json"
+
+
+@pytest.fixture(name="gateway_facade")
+def gateway_facade_fixture(devnet_gateway_url: str):
+    log_color_provider = LogColorProvider()
+    log_color_provider.is_ci_mode = False
```
Comment (mkaput) [inline tests/integration/gateway/gateway_facade_test.py:24]: we have fixture for that

**Review by mkaput — CHANGES_REQUESTED** [review]

File: protostar/starknet_gateway/starknet_request.py (hunk 3 above)
Comment (kasperski95) [reply protostar/starknet_gateway/starknet_request.py]: Resolving: Moved it to the Call Cheatcode PR

File: protostar/starknet_gateway/gateway_facade.py (hunk 2 above)
Comment (kasperski95) [reply protostar/starknet_gateway/gateway_facade.py]: Renamed

File: protostar/commands/init/init_command.py (hunk 4 above)
Comment (kasperski95) [reply protostar/commands/init/init_command.py]: Resolving, removed cwd. Modified `ProtostarFixture::init` instead.

File: tests/integration/gateway/gateway_facade_test.py (hunk 5 above)
Comment (kasperski95) [reply tests/integration/gateway/gateway_facade_test.py]: Resolving: Removed it, because It was longer necessary.

File: protostar/starknet_gateway/gateway_facade.py [hunk 6]
```diff
@@ -18,7 +18,7 @@
 from protostar.starknet_gateway.starknet_request import StarknetRequest
 from protostar.utils.log_color_provider import LogColorProvider
 
-GatewayFacadeSupportedInputType = Union[List[int], Dict[str, Any]]
+ContractFunctionInputType = Union[List[int], Dict[str, Any]]
```
Comment (mkaput) [inline protostar/starknet_gateway/gateway_facade.py:21]: could this type be moved to #618 ?

File: protostar/composition_root.py [hunk 7]
```diff
@@ -152,6 +151,7 @@ def build_di_container(script_root: Path):
             migrator_builder=Migrator.Builder(
                 migrator_execution_environment_builder=MigratorExecutionEnvironment.Builder(),
                 gateway_facade_builder=GatewayFacade.Builder(project_root_path),
+                cwd=project_root_path,
```
Comment (mkaput) [inline protostar/composition_root.py:154]: looks like a dead argument?

**Review by mkaput — CHANGES_REQUESTED** [review]

File: protostar/composition_root.py (hunk 7 above)
Comment (kasperski95) [reply protostar/composition_root.py]: It's not dead. `cwd` was maybe a bad name for it.
Resolving: renamed to `project_root_path`.

File: protostar/starknet_gateway/gateway_facade.py (hunk 6 above)
Comment (kasperski95) [reply protostar/starknet_gateway/gateway_facade.py]: Resolving: Removed type from this PR

**Review by mkaput — APPROVED** [review]
